File: DataStructures/Queue/csharp/Queue.cs
internal class Node<T>
{
    internal T value;
    internal Node<T>? next;

    public Node(T value)
    {
        this.value = value;
        this.next = null;
    }
}

internal class Queue<T>
{
    private Node<T>? firstNode;
    private Node<T>? lastNode;
    private int queueSize;

    public void Enqueue(T value)
    {
        Node<T> curNode = new(value);

        if(queueSize > 0)
        {
            lastNode.next = curNode;
        } else
        {
            firstNode = curNode;
        }

        lastNode = curNode;
        queueSize++;
    }

    public void Dequeue()
    {
        if(queueSize == 0)
        {
            return;
        }

        firstNode = firstNode.next;
        queueSize--;
    }

    public void printQueue()
    {
        var curNode = firstNode;

        while(curNode != null)
        {
            Console.Write($"{curNode.value} ");
            curNode = curNode.next;
        }
        Console.WriteLine("");
    }
}

public class Program
{
    static void Main()
    {
        Queue<int> queue = new();
        queue.Dequeue();
        queue.Enqueue(1);
        queue.Enqueue(3);
        queue.Enqueue(5);
        queue.Enqueue(7);
        queue.Enqueue(9);
        queue.printQueue();
        queue.Dequeue();
        queue.Dequeue();
        queue.printQueue();
    }
    
}

File: DataStructures/Queue/go/queue.go
package main

import "fmt"

type Queue struct {
	len   int
	first *Node
	last  *Node
}

type Node struct {
	value int
	next  *Node
}

func (queue *Queue) Enqueue(value int) {
	if queue.first == nil {
		newNode := &Node{value: value, next: nil}
		queue.first = newNode
		queue.last = newNode

		queue.len++
		return
	}

	lastNode := queue.last
	lastNode.next = &Node{value: value, next: nil}
	queue.last = lastNode.next
	queue.len++
}

func (queue *Queue) Dequeue() {
	if queue.len == 0 {
		return
	}

	deleteNode := queue.first
	queue.first = deleteNode.next
	deleteNode = nil
	queue.len--
}

func (queue *Queue) Print() {
	if queue.len == 0 {
		return
	}

	curNode := queue.first
	fmt.Printf("%d -> ", curNode.value)

	for curNode.next != nil {
		curNode = curNode.next
		fmt.Printf("%d -> ", curNode.value)
	}

	fmt.Println()
}

func main() {
	newQueue := &Queue{}

	newQueue.Enqueue(1)
	newQueue.Enqueue(2)
	newQueue.Enqueue(3)

	newQueue.Print()

	newQueue.Dequeue()

	newQueue.Print()
}
